src/main/scala/com/daml/projection/JdbcAction.scala

```diff
@@ -162,42 +166,69 @@ object Sql {
 }
 
 /**
- * Binds a value to a position in the PreparedStatement.
+ * Binds a value of `T` to a [[java.sql.PreparedStatement]].
  */
-final case class BindValue[T](value: T, pos: Int) extends Setter {
-  def set(ps: java.sql.PreparedStatement) = {
-    value match {
-      case x: Boolean                   => ps.setBoolean(pos, x)
-      case x: Byte                      => ps.setByte(pos, x)
-      case x: Short                     => ps.setShort(pos, x)
-      case x: Int                       => ps.setInt(pos, x)
-      case x: Long                      => ps.setLong(pos, x)
-      case x: Float                     => ps.setFloat(pos, x)
-      case x: Double                    => ps.setDouble(pos, x)
-      case x: BigDecimal                => ps.setBigDecimal(pos, x.bigDecimal)
-      case x: java.math.BigDecimal      => ps.setBigDecimal(pos, x)
-      case x: String                    => ps.setString(pos, x)
-      case x: java.sql.Date             => ps.setDate(pos, x)
-      case x: java.sql.Timestamp        => ps.setTimestamp(pos, x)
-      case x: java.sql.Array            => ps.setArray(pos, x)
-      case null                         => ps.setNull(pos, java.sql.Types.NULL)
-      case x: Optional[_] if x.isEmpty  => ps.setNull(pos, java.sql.Types.NULL)
-      case x: Optional[_] if !x.isEmpty => ps.setObject(pos, x.get())
-      // TODO fix in typeclass solution
-      case x: Option[_] if x.isEmpty => ps.setNull(pos, java.sql.Types.NULL)
-      case Some(x)                   => ps.setObject(pos, x)
-      case x                         => ps.setObject(pos, x)
-    }
-  }
+@FunctionalInterface
+trait Bind[-T] {
+
+  /** Sets the value on the [[java.sql.PreparedStatement]] */
+  def set(ps: PreparedStatement, pos: Int, value: T): Unit
+
+  /** Creates a [[Bind.Applied]] function that will set `value` at `pos` on a `PreparedStatement` argument. */
+  final def apply(value: T, pos: Int): Bind.Applied = set(_, pos, value)
 }
 
-/** Sets a bind value on a PreparedStatement */
-@SerialVersionUID(1L)
-@FunctionalInterface
-trait Setter extends java.io.Serializable {
+/**
+ * Provides `Bind` values for [[ExecuteUpdate.bind[T](pos*]],[[ExecuteUpdate.bind[T](name*]], [[Binder.bind[T](pos*]]
+ * and [[Binder.bind[T](name*]] methods.
+ */
+object Bind {
+  type Used[-T] = Bind[_ >: T]
```

Comment:
I pushed 5edde394852d225196a9540e72b11c2115dc91bf which shows how to make this work. You can read `Used` as a supertype of `Bind`, and generally we should provide `Bind`s but ask for `Used`s, as in that commit.

Reply:
Ah I can see what I missed now, thanks.